### PR TITLE
Write some better algorithms for frame manipulation bits

### PR DESCRIPTION
**Pull request by jgraham**

### CODE REVIEW

File: webdriver-spec.html [hunk 1]
```diff
@@ -1351,24 +1351,31 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
 
+<h2>Command Contexts</h2>
+
+<section>
+  <h3>Introduction</h3>
+
+  <p>Many WebDriver commands happen in the context of either
+    the <a>current browsing context</a> or <a>current top-level browsing
+      context</a>. The <a>current top-level browsing context</a> is
+    represened in the protocol by its associated <a>window
```
Comment (andreastt) [inline webdriver-spec.html]: represented

File: webdriver-spec.html [hunk 2]
```diff
@@ -1351,24 +1351,31 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
- the user agent's operating system level windows through actions such as
- <a rel="Close Window">closing the window</a>,
- <a rel="Set Window Size">resizing</a>,
- <a rel="Maximize Window">maximizing</a>,
- and <a rel="Fullscreen Window">fullscreening</a> a window.
 
+<h2>Command Contexts</h2>
+
+<section>
+  <h3>Introduction</h3>
```
Comment (andreastt) [inline webdriver-spec.html]: Not convinced this needs a header

File: webdriver-spec.html [hunk 3]
```diff
@@ -1351,24 +1351,31 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
+    the <a>current browsing context</a> or <a>current top-level browsing
+      context</a>. The <a>current top-level browsing context</a> is
+    represened in the protocol by its associated <a>window
+      handle</a>. A top-level browsing context is selected using
+    the <a>Switch To Window</a> command. Once this is done, a
+    specific browsing context can be selected using the <a>Switch to
+      Frame</a> command.
+
+  <p class=note>The use of the term "window" to refer to a top-level
+    browsing context is legacy and doesn't correspond with either the
```
Comment (andreastt) [inline webdriver-spec.html]: `<a>top-level browsing context</a>`

File: webdriver-spec.html [hunk 4]
```diff
@@ -1351,24 +1351,31 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
+    operating system notaion of a "window" or the DOM Window object.</p>
```
Comment (andreastt) [inline webdriver-spec.html]: Link to DOM Window?

File: webdriver-spec.html [hunk 5]
```diff
@@ -1351,24 +1351,32 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
+
+  <p class=note>The use of the term "window" to refer to a top-level
+    browsing context is legacy and doesn't correspond with either the
+    operating system notaion of a "window" or the
+    DOM <a href="https://html.spec.whatwg.org/#the-window-object">Window</a>
+    object.</p>
+
+  <p class=note>In accordance with
+    the <a href="http://www.w3.org/TR/html5/editing.html#focus">focus</a>
+    section of the [[!html51] spec, commands are unaffected by whether
```
Comment (andreastt) [inline webdriver-spec.html]: Missing `]`

File: webdriver-spec.html [hunk 6]
```diff
@@ -1351,24 +1351,32 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
+    the operating-system window has focus or not.</p>
```
Comment (andreastt) [inline webdriver-spec.html]: operating system, without dash

File: webdriver-spec.html [hunk 7]
```diff
@@ -1516,7 +1524,119 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
+         <th>Notes</th>
+       </tr>
+       <tr>
+         <td>POST</td>
+         <td id='post-frame'>/session/{sessionId}/frame</td>
+         <td></td>
+       </tr>
+     </table>
+     <p>The <a>Switch to Frame</a> command is used to select a child
+     browsing context of the <a>current browsing context</a> to use as
```
Comment (andreastt) [inline webdriver-spec.html]: Link _child browsing context_?

File: webdriver-spec.html [hunk 8]
```diff
@@ -1516,7 +1524,119 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
+     the <a>current browsing context</a> for subsequent commands.</p>
+     <p>The <a>remote end steps</a> for the <dfn>Switch to Frame</dfn>
+     command are:</p>
+     <ol>
+
+       <li><p>If the <a>current browsing context</a> is <a>no longer open</a>,
+           return an error with code <a>no such window</a>.</li>
+
+       <li><p>Let <var>id</var> be the result of <a>getting a property</a> named "id"
```
Comment (andreastt) [inline webdriver-spec.html]: `"<code>id</code>"`

File: webdriver-spec.html [hunk 9]
```diff
@@ -1516,7 +1524,119 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
+           from the <var>parameters</var> argument.</li>
+
+       <li>
+         <p>If <var>id</var> is null:
```
Comment (andreastt) [inline webdriver-spec.html]: Feel we could use a switch style statement that many other specs use here…  But this is good for now.

File: webdriver-spec.html [hunk 10]
```diff
@@ -1516,7 +1524,119 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
+           <ol><li><p>Set the <a>current browsing context</a> to
+                 the <a>current top-level browsing context</a>.</li>
+           </ol>
+         <p>Otherwise if <var>id</var> is a <code>Number</code> object:</p>
+         <ol>
+           <li><p>If <var>id</var> is less than 0 or greater than
+           2<sup>16</sup> - 1, return <a>error</a> with status <a>no
```
Comment (andreastt) [inline webdriver-spec.html]: Why this number specifically?

File: webdriver-spec.html [hunk 11]
```diff
@@ -1516,7 +1524,119 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
+           such frame</a>.</li>
+           <li><p>Let <var>window</var> be the
+               <a href="https://html.spec.whatwg.org/#concept-document-window">associated
+                 window</a> of the <a>current browsing
+                 context</a>'s <a href="https://html.spec.whatwg.org/#active-document">active
+                 document.</a></li>
+           <li><p>If <var>id</var> is not
+               a <a href="http://heycam.github.io/webidl/#dfn-supported-property-indices">supported
+                 property index</a> of <var>window</var>,
+               return <a>error</a> with status <a>no such frame</a></li>
```
Comment (andreastt) [inline webdriver-spec.html]: Missing full stop.

File: webdriver-spec.html [hunk 12]
```diff
@@ -1516,7 +1524,119 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
+           <li><p>Let <var>child window</var> be the <code>WindowProxy</code> object
+               obtained
+               by <a href="https://html.spec.whatwg.org/#dom-window-item">determining
+                 the value of an indexed property</a> of <var>window</var>
+               with index <var>id</var>.</li>
+           <li><p>Set the <a>current browsing context</a> to <var>new
+           window</var>'s browsing context.</li>
+         </ol>
+         <p>Otherwise, if <var>id</var> <a>represents a WebElement</a>:
```
Comment (andreastt) [inline webdriver-spec.html]: <a>web element</a>

File: webdriver-spec.html [hunk 13]
```diff
@@ -1516,7 +1524,119 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
+             <a href="https://html.spec.whatwg.org/#nested-browsing-context">nested
+             browsing context</a>.</li>
+           </ol>
+           <p>Otherwise:
+             <ol>
+               <li><p>Return <a>error</a> with status <a>no such
+               frame</a>.</li>
+             </ol>
+       </li>
+       <li><p>Return <a>success</a> with data null</li>
```
Comment (andreastt) [inline webdriver-spec.html]: Missing full stop

File: webdriver-spec.html [hunk 14]
```diff
@@ -1516,7 +1524,119 @@ <h3 id="page-load-strategies">The Page Load Strategy</h3>
...
+     <ol>
+       <li><p>If the <a>current browsing context</a> is <a>no longer open</a>,
+           return an error with code <a>no such window</a>.</li>
+       <li><p>If the <a>current browsing context</a> is not equal to
+       the <a>current top-level browsing context</a>, set
+       the <a>current browsing context</a> to
+       the <a href="https://html.spec.whatwg.org/#parent-browsing-context">parent
+       browsing context</a> of the <a>current browsing
+       context</a>.</li>
+       <li><p>Return <a>success</a> with data null</li>
```
Comment (andreastt) [inline webdriver-spec.html]: Missing full stop.

**Comment by andreastt** [comment]
This looks good to me, with some fixups and a rebase.

**Comment by andreastt** [comment]
👍🏻

File: webdriver-spec.html (hunk 10 above)
Comment (jgraham) [inline webdriver-spec.html]: No idea, I was just copying what the IDL used to say about this being a short. It doesn't seem particularly sane to me to have this restriction.

File: webdriver-spec.html (hunk 12 above)
Comment (jgraham) [inline webdriver-spec.html]: The idea is that `represents a web element` will be a defined thing.

**Comment by AutomatedTester** [comment]
landed in 8d1e80b

**Pull request closed without merging**